### PR TITLE
blog post about pyodide web app that uses pandas and other python packages

### LINKED ISSUE #46: Post Idea: Visual low/no code data preparation and analysis app built with pyodide and react
I built a proof of concept app with pyodide that allows users to use the pandas library without needing to code. I also included functionality for plotting and linear regression. I would like to create a post discussing the project and why someone may benefit from it. Thank you.

You can read more about it [here](https://github.com/randr000/react_pyodide_data_prep)

You can test it out [here](https://randr000.github.io/react_pyodide_data_prep/)

pyodide

### PR DESCRIPTION
**Pull request by randr000**
blog post about pyodide web app that uses pandas and other python packages

### CODE REVIEW

**Review by ryanking13 — COMMENTED** [review]
Thanks for writing the post @randr000! Would you mind adding some more detail about your project? Such as:

- Some implementation details
- Pain points while implementing the app
- The parts you liked or disliked about Pyodide.

**Comment by randr000** [comment]
Updated.

File: content/posts/pyodide-react-pandas-matplotlib-scikit.md [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    # caption: "<text>" # display caption under cover
+    relative: false # when using page bundles set this to true
+    hidden: true # only hide on current single page
+---
+
+Hello, my name is [Raul Andrial](https://randr000.github.io/portfolio-resume/tech) and I am a software engineer located in Miami, FL, USA. I created a proof of concept web app with Pyodide that allows users to use the pandas library without needing to code but can use code if they want to. The app uses a drag and drop interface. I also included functionality for plotting with matplotlib and linear regression using scikit-learn.
+
+You can read more about it [here](https://github.com/randr000/react_Pyodide_data_prep).
+
+You can test it out [here](https://randr000.github.io/react_Pyodide_data_prep/).
```
Comment (ryanking13) [inline content/posts/pyodide-react-pandas-matplotlib-scikit.md:34]: This link seems broken.

File: content/posts/pyodide-react-pandas-matplotlib-scikit.md [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+## Pyodide Context Provider
+
+The Pyodide instance is loaded and stored in its own context using React's [Context API](https://react.dev/reference/react/createContext). In order for any component to run Python code it must be a child of the PyodideContextProvider component.
+
+```
```
Comment (ryanking13) [inline content/posts/pyodide-react-pandas-matplotlib-scikit.md:40]:
```suggestion
```jsx
```

For better rendering

File: content/posts/pyodide-react-pandas-matplotlib-scikit.md [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+export const PyodideContextProvider = ({children, toLoadPyodide=true}) => {
+
+    const [isPyodideLoaded, setIsPyodideLoaded] = useState(!toLoadPyodide);
+    const PyodideRef = useRef(null);
+
+    useEffect(() => {
+        toLoadPyodide &&
+        (async function () {
+            PyodideRef.current = await loadPyodide({
+                indexURL : "https://cdn.jsdelivr.net/Pyodide/v0.21.3/full/"
```
Comment (ryanking13) [inline content/posts/pyodide-react-pandas-matplotlib-scikit.md:55]:
```suggestion
                indexURL : "https://cdn.jsdelivr.net/pyodide/v0.21.3/full/"
```

File: content/posts/pyodide-react-pandas-matplotlib-scikit.md [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+            });
+
+            // Load Python packages
+            ['pandas', 'numpy', 'matplotlib', 'scikit-learn', 'cloudpickle'].forEach(async pkg => await PyodideRef.current.loadPackage(pkg));
```
Comment (ryanking13) [inline content/posts/pyodide-react-pandas-matplotlib-scikit.md:59]: you don't need to pass packages one-by-one. `pyodide.loadPackage('pandas', 'numpy', 'matplotlib', 'scikit-learn', 'cloudpickle']`) will do the job.

File: content/posts/pyodide-react-pandas-matplotlib-scikit.md (hunk 3 above)
Comment (ryanking13) [reply content/posts/pyodide-react-pandas-matplotlib-scikit.md:55]: Just curious, why did you use Pyodide v0.21.3, which was released quite a while ago? Was there some issue using more recent versions of Pyodide?

File: content/posts/pyodide-react-pandas-matplotlib-scikit.md [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+```
+-- Pyodide: The current instance of Pyodide.
+<br/>
+-- isPyodideLoaded: A boolean value that can be used to check if Pyodide has been loaded.
+
+## Defining Python Functions
+
+The way I run the Python code is by defining Python functions within a javascript template literal string. Below is an example of a function that adds two numbers.
+
+```
```
Comment (ryanking13) [inline content/posts/pyodide-react-pandas-matplotlib-scikit.md:96]:
```suggestion
```js
```

File: content/posts/pyodide-react-pandas-matplotlib-scikit.md [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+def addTwoNums(x, y):
+    return x + y`;
+
+export default addTwoNums;
+```
+## Running Python Code
+In order to run Python code, I import a previously defined function and use the useGetContexts hook to reference the current Pyodide instance. I then call the runPython method to load the Python function and call globals.get to call the Python function. This usually happens within a useEffect hook. Below is an example using the addTwoNums function defined above.
+
+```
```
Comment (ryanking13) [inline content/posts/pyodide-react-pandas-matplotlib-scikit.md:107]:
```suggestion
```jsx
```

**Review by ryanking13 — COMMENTED** [review]
Thanks! I left some minor comments about readability. The content looks good to me.

File: content/posts/pyodide-react-pandas-matplotlib-scikit.md (hunk 2 above)
Comment (randr000) [reply content/posts/pyodide-react-pandas-matplotlib-scikit.md]: Changed.

File: content/posts/pyodide-react-pandas-matplotlib-scikit.md (hunk 3 above)
Comment (randr000) [reply content/posts/pyodide-react-pandas-matplotlib-scikit.md]: Changed. I had started this project back in 2022 and then didn't touch it for about a year. I tried using a more recent version when I revisited this project this year, but had issues with importing the pyodide package in one of my python scripts so I reverted to using an old version for now and plan on using a newer version later when I revisit again.

File: content/posts/pyodide-react-pandas-matplotlib-scikit.md (hunk 4 above)
Comment (randr000) [reply content/posts/pyodide-react-pandas-matplotlib-scikit.md]: Changed.

File: content/posts/pyodide-react-pandas-matplotlib-scikit.md (hunk 5 above)
Comment (randr000) [reply content/posts/pyodide-react-pandas-matplotlib-scikit.md]: Changed.

File: content/posts/pyodide-react-pandas-matplotlib-scikit.md (hunk 6 above)
Comment (randr000) [reply content/posts/pyodide-react-pandas-matplotlib-scikit.md]: Changed.

**Review by ryanking13 — APPROVED** [review]
It looks good to me. @hoodmane do you want to add more comments?

**Comment by randr000** [comment]
Are there any more issues that I need to address?

**Comment by ryanking13** [comment]
> Are there any more issues that I need to address?

No, sorry for the delay, let me update the publish date and release.

**Pull request merged**